pkg/gardenlet/controller/federatedseed/shootstate/common_sync.go

```diff
@@ -0,0 +1,155 @@
+// limitations under the License.
+
+package shootstate
+
+import (
+	"fmt"
+	"time"
+
+	apiextensions "github.com/gardener/gardener/pkg/api/extensions"
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	gardencorev1alpha1helper "github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+	extensionsclient "github.com/gardener/gardener/pkg/client/extensions/clientset/versioned"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var decoder runtime.Decoder
+
+func init() {
+	scheme := runtime.NewScheme()
+	_ = gardencorev1alpha1.AddToScheme(scheme)
+
+	decoder = serializer.NewCodecFactory(scheme).UniversalDecoder()
+}
+
+func createEnqueueOnAddFunc(queue workqueue.RateLimitingInterface) func(extensionObject interface{}) {
+	return func(newObj interface{}) {
+		extensionObject := newObj.(extensionsv1alpha1.Object)
+		if extensionObject.GetExtensionStatus().GetState() == nil {
+			return
+		}
+		enqueue(queue, newObj)
+	}
+}
+
+func createEnqueueOnUpdateFunc(queue workqueue.RateLimitingInterface) func(newExtensionObject, oldExtensionObject interface{}) {
+	return func(newObj, oldObj interface{}) {
+		var (
+			newExtensionObj = newObj.(extensionsv1alpha1.Object)
+			oldExtensionObj = oldObj.(extensionsv1alpha1.Object)
+		)
+
+		if newExtensionObj.GetExtensionStatus().GetState() == nil {
+			return
+		}
+
+		if oldExtensionObj.GetExtensionStatus().GetState() == nil {
+			enqueue(queue, newObj)
+		}
+
+		if apiequality.Semantic.DeepEqual(newExtensionObj.GetExtensionStatus().GetState().Raw, oldExtensionObj.GetExtensionStatus().GetState().Raw) {
+			return
+		}
+		enqueue(queue, newObj)
+	}
+}
+
+func enqueue(queue workqueue.RateLimitingInterface, obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	queue.Add(key)
+}
+
+func (s *SyncController) createShootStateSyncReconcileFunc(kind string, obj runtime.Object) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		err := s.seedClient.Client().Get(s.context, req.NamespacedName, obj)
+		if apierrors.IsNotFound(err) {
+			s.log.Debug("Skipping because resoruce with kind was deleted")
+			return reconcile.Result{}, nil
+		}
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		extensionObject, err := apiextensions.Accessor(obj)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+
+		state := extensionObject.GetExtensionStatus().GetState()
+		if state == nil {
```

Comment:
What's the motivation for this additional check? Cannot happen, can it? And even if, what's the problem with it?

Reply:
If state is nil we should just skip this syncing it.

state = nil means that the the extension resource does not have to preserve its state.

Reply:
In a very corner case where the State is being deleted between adding the object to the queue and the workers processing it. The thing is if that happens it will result in a runtime panic error.